// @waylay/client/lib/byoml.js
'use strict'

const Resource = require('./resource')

/**
 * @namespace Byoml
 */
class Byoml extends Resource {
  /**
   * List your models
   * @return {Promise<Model[]>} A list of models
   */
  list () {
    return this._request({ path: '/models' })
  }

  /**
   * Get information about a model
   * @param  {string} name The name of the model
   * @return {Promise<Model>}
   */
  get (name) {
    return this._request({
      path: `/models/${name}`
    })
  }

  /**
   * Get an example predict payload for a model
   * @param  {string} name The name of the model model
   * @return {Promise} An example payload object
   */
  exampleInput (name) {
    return this._request({
      path: `/models/${name}/examples`
    })
  }

  /**
   * Infer predictions from a model
   * See {@link https://docs.waylay.io/#/api/byoml/?id=predictions}
   * @param  {string} name The name of the model
   * @param  {Object} payload The payload to infer predictions
   * @return {Promise} Object containing the predictions
   */
  predict (name, payload) {
    return this._request({
      path: `/models/${name}/predict`,
      method: 'POST',
      data: payload
    })
  }

  /**
   * Infer predictions from a model using regression
   * @param  {string} name The name of the model
   * @param  {Object} payload The payload to infer
   * @return {Promise} Object containing the predictions
   */
  regress (name, payload) {
    return this._request({
      path: `/models/${name}/regress`,
      method: 'POST',
      data: payload
    })
  }

  /**
   * Infer predictions from a model using classification
   * @param  {string} name The name of the model
   * @param  {Object} payload The payload to infer
   * @return {Promise} Object containing the predictions
   */
  classify (name, payload) {
    return this._request({
      path: `/models/${name}/classify`,
      method: 'POST',
      data: payload
    })
  }

  /**
   * Remove a model
   * @param  {string} name The name of the model
   * @return {Promise}
   */
  remove (name) {
    return this._request({
      method: 'DELETE',
      path: `/models/${name}`
    })
  }
}

module.exports = Byoml
