// @waylay/client/lib/policies.js
'use strict'

const Resource = require('./resource')

/**
 * @namespace Policies
 */
class Policies extends Resource {
  /**
   * List all policies
   *
   * @return {Promise<Policy[]>} A list of policies
   */
  list () {
    return this._request({ path: '/' })
  }

  /**
   * Get details of one individual policy
   *
   * @param  {string} id
   * @return {Promise<Policy>}
   */
  get (id) {
    return this._request({ path: `/${id}` })
  }

  /**
   * Create a new policy
   *
   * @param  {Policy} policy The payload to create a new policy
   * @return {Promise<Policy>}
   */
  create (policy) {
    return this._request({
      path: '/',
      method: 'POST',
      data: policy
    })
  }

  /**
   * Update a policy
   *
   * @param  {string} id
   * @return {Promise<Policy>}
   */
  update (id, policy) {
    return this._request({
      path: `/${id}`,
      method: 'PUT',
      data: policy
    })
  }

  /**
   * Update a policy, allows partial updates
   * @param  {string} id
   * @param  {Object} policy
   * @return {Promise<Policy>}
   */
  patch (id, policy) {
    return this._request({
      path: `/${id}`,
      method: 'PATCH',
      data: policy
    })
  }

  /**
   * Remove a policy
   *
   * @param  {string} id
   * @return {Promise}
   */
  remove (id) {
    return this._request({
      path: `/${id}`,
      method: 'DELETE'
    })
  }
}

module.exports = Policies
