// @waylay/client/lib/auth/twoFA.js
const Resource = require('../resource')

/**
 * @namespace TwoFA
 */
class TwoFA extends Resource {
  /**
   * Fetch the 2FA status for current token
   * Optionaly, applications or sudo users can pass a user & tenant combination
   * @param  {string} user
   * @param  {string} tenant
   * @return {Promise}
   */
  status (user, tenant) {
    // Only include values in params that are not empty
    const params = Object.fromEntries([
      ['userId', user],
      ['tenantId', tenant]
    ].filter(i => i[1]))

    return this._request({
      path: `/auth/2fa/status`,
      params
    })
  }

  /**
   * Enable 2FA enforcement on a tenant
   * @param  {string} tenant
   * @return {Promise}
   */
  enableTwoFAEnforcement (tenant) {
    return this._request({
      path: `/auth/2fa/${tenant}/enforce`,
      method: 'POST',
      data: { state: true }
    })
  }

  /**
   * Disable 2FA enforcement on a tenant
   * @param  {string} tenant
   * @return {Promise}
   */
  disableTwoFAEnforcement (tenant) {
    return this._request({
      path: `/auth/2fa/${tenant}/enforce`,
      method: 'POST',
      data: { state: false }
    })
  }
}

module.exports = TwoFA
